server/src/main/java/org/opensearch/plugins/TelemetryAwarePlugin.java

```diff
@@ -0,0 +1,80 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.client.Client;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.common.lifecycle.LifecycleComponent;
+import org.opensearch.core.common.io.stream.NamedWriteable;
+import org.opensearch.core.common.io.stream.NamedWriteableRegistry;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.env.Environment;
+import org.opensearch.env.NodeEnvironment;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.script.ScriptService;
+import org.opensearch.telemetry.metrics.MetricsRegistry;
+import org.opensearch.telemetry.tracing.Tracer;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.watcher.ResourceWatcherService;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.function.Supplier;
+
+/**
+ * Plugin that provides the telemetry registries to build component with telemetry and also provide a way to
+ * pass telemetry registries to the implementing plugins for adding instrumentation in the code.
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public interface TelemetryAwarePlugin {
+
+    /**
+     * Returns components added by this plugin.
+     * <p>
+     * Any components returned that implement {@link LifecycleComponent} will have their lifecycle managed.
+     * Note: To aid in the migration away from guice, all objects returned as components will be bound in guice
+     * to themselves.
+     *
+     * @param client A client to make requests to the system
+     * @param clusterService A service to allow watching and updating cluster state
+     * @param threadPool A service to allow retrieving an executor to run an async action
+     * @param resourceWatcherService A service to watch for changes to node local files
+     * @param scriptService A service to allow running scripts on the local node
+     * @param xContentRegistry the registry for extensible xContent parsing
+     * @param environment the environment for path and setting configurations
+     * @param nodeEnvironment the node environment used coordinate access to the data paths
+     * @param namedWriteableRegistry the registry for {@link NamedWriteable} object parsing
+     * @param indexNameExpressionResolver A service that resolves expression to index and alias names
+     * @param repositoriesServiceSupplier A supplier for the service that manages snapshot repositories; will return null when this method
+     *                                   is called, but will return the repositories service once the node is initialized.
+     * @param tracer the tracer to add tracing instrumentation.
+     * @param metricsRegistry the registry for metrics instrumentation.
+     */
+    default Collection<Object> createComponents(
```

Comment:
@reta and I chatted about this and I agree this works and makes the initialization lifecycle clear. The general concern is that if we continue to encounter situations where additional data needs to be passed to certain plugins, we'll have to repeat this pattern which seems like it could get out of hand (I don't know for sure if this will be a problem though...). I just want to throw out another option that might work which is creating the concept of a "post initialization callback" that is guaranteed to happen after createComponents has been called on all plugins. Something like the following can be added to `org.opensearch.plugins.Plugin` (or maybe just a new class?):

```java
/**
 * Callback that is invoked after all plugins have been initialized, providing a hook
 * for the system to provide instances and data that they themselves may have been
 * created by other plugins. This call is guaranteed to happen after createComponents.
 */
public interface PostInitializationCallback {
    default void onTelemetryInitialized(Tracer tracer, MetricsRegistry metricsRegistry) {}
}

public PostInitializationCallback getInitializationCallback() {
    // default no-op implementation. Plugins can provide their
    // own implementation to receive the data passed into the callback.
    return new PostInitializationCallback() {};
}
```

The huge caveat here is that these instances would not be available for a Plugin to use in createComponents() (not sure if that's a requirement). @reta what do you think? Would this be a more extensible pattern?

I'm definitely not going to block what is introduced here if my idea doesn't work.

Reply:
> The huge caveat here is that these instances would not be available for a Plugin to use in createComponents() (not sure if that's a requirement). @reta what do you think?

This is basically the general concern I have with every other approach: we don't really know what exactly telemetry plugin might need for initialization, so having `createComponents` overloaded method is safe bet here - we provide everything we used to provide + a bit more.  More on the subject, it is very, very likely that the components plugins would create, would need the instance of Tracer and / or MetricsRegistry but splitting those apart forces to look for workarounds.

Also, the `PostInitializationCallback` sounds very abstract. We really have a `TelemetryPlugin`, which we use as a provider for telemetry, and `TelemetryAwarePlugin` is a logical continuation of that, very coupled with `TelemetryPlugin`.

Reply:
I do agree though looking into extensible initialization mechanism is needed, but it would be the effort of different scale.

Reply:
> it is very, very likely that the components plugins would create, would need the instance of Tracer and / or MetricsRegistry but splitting those apart forces to look for workarounds.

That's fair and basically makes my idea a non-starter.

> Also, the PostInitializationCallback sounds very abstract. We really have a TelemetryPlugin, which we use as a provider for telemetry, and TelemetryAwarePlugin is a logical continuation of that, very coupled with TelemetryPlugin.

I'm going to nitpick here, but I would argue that "telemetry" is a core concept that this new plugin is exposing to plugins for _usage_. It just so happens that part of the telemetry implementation itself is pluggable, but plugins that are "telemetry aware" just want to use the feature, and unlike TelemetryPlugin they are unconcerned with the details of how to implement telemetry. There is obviously a strict ordering requirement in terms of initialization, but no actual logical coupling of "TelemetryPlugin" and "TelemetryAwarePlugin". But none of that really matters because I think our framework isn't really designed to handle this initialization ordering requirement in a good way, so this solution is probably the best way to proceed now.

Reply:
> I'm going to nitpick here, but I would argue that "telemetry" is a core concept that this new plugin is exposing to plugins for usage. It just so happens that part of the telemetry implementation itself is pluggable, but plugins that are "telemetry aware" just want to use the feature, and unlike TelemetryPlugin they are unconcerned with the details of how to implement telemetry. 

Thanks @andrross , I agree with the conceptually, but `TelemetryPlugin` is a plugin, the APIs are in core. And because of that the initialization becomes non-intuitive: we treat `TelemetryPlugin` in a special way, by initializing it before every other plugin. `TelemetryAwarePlugin` highlights this special treatment.